pandas/plotting/_matplotlib/timeseries.py

```diff
@@ -243,7 +243,8 @@ def maybe_convert_index(ax: "Axes", data):
 
         if freq is None:
             # We only get here for DatetimeIndex
```

Comment:
we might need an additional cast and a comment that freq is never None for PeriodIndex

The mypy errors are 
```
pandas\plotting\_matplotlib\timeseries.py:247: error: Item "DatetimeIndex" of "Union[DatetimeIndex, PeriodIndex]" has no attribute "inferred_freq"  [union-attr]
pandas\plotting\_matplotlib\timeseries.py:247: error: Item "PeriodIndex" of "Union[DatetimeIndex, PeriodIndex]" has no attribute "inferred_freq"  [union-attr]
```

so once the issue with the dynamically added `inferred_freq` in `DatetimeIndex` is sorted, this will still error.

Reply:
Thanks to your suggestion below, neither are needed. mypy correctly infers that `data.index` is a DatetimeIndex when freq is None.

Reply:
Minor update: mypy would be able to do the inference above, except that PeriodIndex.freq (a BaseOffset) is typed as Any since BaseOffset is a cdef.